### DISCUSSION TITLE
How to access data you'd store in cookies or localStorage on server side?

### DISCUSSION THREAD

**Discussion by tjdam** [vercel/next.js#17247] [Help] [answered]
Hi!

I use JWT to control the access to certain fields in GraphQL queries, which works amazingly for queries within components using Apollo's useQuery and getting the token from cookies. But how can one store this kind of information on the server side?

Using getStaticProps or getServerSideProps you have absolutely no access to that. I thought of storing it in memory, but I don't know when exactly I get access to what goes in memory server side.

Could anyone help me understand this?

Thanks :)

**Comment by jamesmosier** [comment, marked as answer]
As you mentioned `localStorage` and `sessionStorage` are not available on the server. You could use cookie and access them inside `getServerSideProps` via `req.cookies`. [See this example](https://github.com/vercel/next.js/blob/31ab12aaf779ba2d429d6093cf94281df40e9771/examples/with-passport/lib/auth-cookies.js#L28) for some tips on working with cookies and Next. There are also a few other [cookie examples](https://github.com/vercel/next.js/tree/canary/examples/with-cookie-auth) you could take a look at.

You won't be able to use this with `getStaticProps` or `getStaticPaths` though because this is run at build time and therefore does not know anything about the users browser at that time.

**Reply by tjdam** [reply]
Ohhh that makes a lot of sense! Now it seems obvious, but thanks for that, it really made things a lot more clear for me now. I have so many issues understanding when something is happening in the server or in the client.

But that being said, shouldn't I be able to have access to this req.cookies in the getStaticProps if I have fallback as true in the getStaticPaths? Doesn't it behave like a SSR in the case it finds a path that is not defined?

Thanks!

**Reply by jamesmosier** [reply]
You won't be able to access cookies via `getStaticProps` or `getStaticPaths`, even with `fallback: true`. The primary purpose of those functions are for static generation during build and cookies could never be available there, so (I'm guessing) the functionality was not built in to do that because for most use cases it would never be able to use cookies. Also even with `fallback: true` whenever the first user hits that page then it would be built and then that same copy would be served to all other users.

**Reply by tjdam** [reply]
Thanks so much for your time :)

**Reply by kirasiris** [reply]
@jamesmosier a bit late to the party? Do I need to set the cookie with setHeader?

I currently have my function returning a JSON response: [carbon.now.sh](https://carbon.now.sh/?bg=rgba%28171%2C+184%2C+195%2C+1%29&t=seti&wt=none&l=auto&ds=true&dsyoff=20px&dsblur=68px&wc=true&wa=true&pv=56px&ph=56px&ln=false&fl=1&fm=Hack&fs=14px&lh=133%25&si=false&es=2x&wm=false&code=const%2520sendTokenResponse%2520%253D%2520%28user%252C%2520statusCode%252C%2520req%252C%2520res%29%2520%253D%253E%2520%257B%250A%2520%2520%252F%252F%2520Create%2520token%250A%2520%2520const%2520token%2520%253D%2520user.getSignedJwtToken%28%29%253B%250A%250A%2520%2520%252F%252F%2520One%2520day%2520equals%2520to...%250A%2520%2520const%2520daysInTime%2520%253D%252024%2520*%252060%2520*%252060%2520*%25201000%253B%250A%250A%2520%2520%252F%252F%2520To%2520work%2520with%2520ReactJS%250A%2520%2520res%250A%2520%2520%2520%2520.status%28statusCode%29%250A%2520%2520%2520%2520.cookie%28%27xAuthToken%27%252C%2520token%252C%2520%257B%250A%2520%2520%2520%2520%2520%2520expires%253A%2520new%2520Date%28%250A%2520%2520%2520%2520%2520%2520%2520%2520Date.now%28%29%2520%252B%2520process.env.JWT_COOKIE_EXPIRE%2520*%2520daysInTime%250A%2520%2520%2520%2520%2520%2520%29%252C%250A%2520%2520%2520%2520%2520%2520httpOnly%253A%2520true%252C%250A%2520%2520%2520%2520%2520%2520secure%253A%2520process.env.NODE_ENV%2520%21%253D%253D%2520%27development%27%250A%2520%2520%2520%2520%257D%29%250A%2520%2520%2520%2520.json%28%257B%2520success%253A%2520true%252C%2520token%2520%257D%29%253B%250A%257D%253B) snippet? 

I would like to mention that I'm using an external API built with Express.

**Reply by hossam1231** [reply]
fast forward 2023 you can use headers from next/headers
